README.md
[![MIT license](http://img.shields.io/badge/license-MIT-brightgreen.svg)](http://opensource.org/licenses/MIT)
[![Latest Stable Version](https://poser.pugx.org/cheprasov/php-redis-client/v/stable)](https://packagist.org/packages/cheprasov/php-redis-client)
[![Total Downloads](https://poser.pugx.org/cheprasov/php-redis-client/downloads)](https://packagist.org/packages/cheprasov/php-redis-client)
# RedisClient v1.10.0 for PHP >= 5.5

## About
RedisClient is a fast, fully-functional and user-friendly client for Redis, optimized for performance. RedisClient supports the latest versions of Redis starting from __2.6__ to __6.0__

## Main features
- Support Redis versions from __2.6.x__ to __6.0.x__.
- Support __TCP/IP__ and __UNIX__ sockets.
- Support __PubSub__ and __Monitor__ functionallity.
- Support __Pipeline__ and __Transactions__.
- Support __Redis Cluster__.
- Support __RAW__ commands as arrays `['SET', 'foo', 'bar']`.
- Connections to Redis are established lazily by the client upon the first command.
- Easy to use with IDE, client has PHPDocs for all supported versions.
- By default, the client works with the latest stable version of Redis (6.0).
- The client was tested on the next latest versions of Redis: `6.0.5`, `5.0.5`, `4.0.14`, `3.2.8`, `3.0.7`, `2.8.24`, `2.6.17`.
- Also, the client was tested on PHP `7.4`, `7.3`, `7.2`, `7.1`, `7.0`, `5.6`, `5.5`, `HHVM`.

## Redis Commands

Please see list of commands here [./COMMANDS.md](https://github.com/cheprasov/php-redis-client/tree/master/COMMANDS.md)

## Usage

### Config

```php
$config = [
    // Optional. Default = '127.0.0.1:6379'. You can use 'unix:///tmp/redis.sock'
    'server' => '127.0.0.1:6379',

    // Optional. Default = 1
    // The timeout for reading/writing data over the socket
    'timeout' => 2,

    // Optional. Default = null
    // See more here: http://php.net/manual/en/function.stream-socket-client.php
    'connection' => [
        // Optional. Default = ini_get("default_socket_timeout")
        // The timeout only applies while making connecting the socket
        'timeout' => 2,

        // Optional. Default = STREAM_CLIENT_CONNECT
        // Bitmask field which may be set to any combination of connection flags.
        // Currently the select of connection flags is limited to STREAM_CLIENT_CONNECT (default),
        // STREAM_CLIENT_ASYNC_CONNECT and STREAM_CLIENT_PERSISTENT.
        'flags' => STREAM_CLIENT_CONNECT
    ],

    // Optional. Specify version to avoid some unexpected errors.
    'version' => '4.0.10',

    // Optional. Use it only if Redis server requires password (AUTH)
    'password' => 'some-password',

    // Optional. Use it, if you want to select not default db (db != 0) on connect
    'database' => 1,

    // Optional. Array with configs for RedisCluster support
    'cluster' => [
        'enabled' => false,

        // Optional. Default = []. Map of cluster slots and servers
        // array(max_slot => server [, ...])
        // Examples for Cluster with 3 Nodes:
        'clusters' => [
            5460  => '127.0.0.1:7001', // slots from 0 to 5460
            10922 => '127.0.0.1:7002', // slots from 5461 to 10922
            16383 => '127.0.0.1:7003', // slots from 10923 to 16383
        ],

        // Optional. Default = false.
        // Use the param to update cluster slot map below on init RedisClient.
        // RedisClient will execute command CLUSTER SLOTS to get map.
        'init_on_start' => false,

        // Optional. Default = false.
        // If Redis returns error -MOVED then RedisClient will execute
        // command CLUSTER SLOTS to update cluster slot map
        'init_on_error_moved' => true,

        // Optional. Defatult = 0.05 sec. It is timeout before next attempt on TRYAGAIN error.
        'timeout_on_error_tryagain' => 0.25, // sec
    ]
];
```

### Create a new instance of RedisClient
```php
<?php
namespace Examples;

require (dirname(__DIR__).'/vendor/autoload.php');
// or require (dirname(__DIR__).'/src/autoloader.php');

use RedisClient\RedisClient;
use RedisClient\Client\Version\RedisClient2x6;
use RedisClient\ClientFactory;

// Example 1. Create new Instance for Redis version 6.0.x with config via factory
$Redis = ClientFactory::create([
    'server' => '127.0.0.1:6379', // or 'unix:///tmp/redis.sock'
    'timeout' => 2,
    'version' => '6.0'
]);

echo 'RedisClient: '. $Redis->getSupportedVersion() . PHP_EOL; // RedisClient: 2.8


// Example 2. Create new Instance without config. Client will use default config.
$Redis = new RedisClient();
// By default, the client works with the latest stable version of Redis.
echo 'RedisClient: '. $Redis->getSupportedVersion() . PHP_EOL; // RedisClient: 3.2
echo 'Redis: '. $Redis->info('Server')['redis_version'] . PHP_EOL; // Redis: 3.0.3


// Example 3. Create new Instance with config
// By default, the client works with the latest stable version of Redis.
$Redis = new RedisClient([
    'server' => '127.0.0.1:6387', // or 'unix:///tmp/redis.sock'
    'timeout' => 2
]);

echo 'RedisClient: '. $Redis->getSupportedVersion() . PHP_EOL; // RedisClient: 3.2
echo 'Redis: '. $Redis->info('Server')['redis_version'] . PHP_EOL; // Redis: 3.2.0


// Example 4. Create new Instance for Redis version 2.6.x with config
$Redis = new RedisClient2x6([
    'server' => 'tcp://127.0.0.1:6379', // or 'unix:///tmp/redis.sock'
    'timeout' => 2
]);

echo 'RedisClient: '. $Redis->getSupportedVersion() . PHP_EOL; // RedisClient: 2.6
```

### Example
Please, see examples here: https://github.com/cheprasov/php-redis-client/tree/master/examples

- [Create new instance](https://github.com/cheprasov/php-redis-client/tree/master/examples/create_new_instance.php)
- [Using MONITOR](https://github.com/cheprasov/php-redis-client/tree/master/examples/monitor.php)
- [Publish and Subscribe](https://github.com/cheprasov/php-redis-client/tree/master/examples/pubsub.php)
- [Transactions](https://github.com/cheprasov/php-redis-client/tree/master/examples/transactions.php)
- [Pipeline](https://github.com/cheprasov/php-redis-client/tree/master/examples/pipeline.php)
- [Cluster support](https://github.com/cheprasov/php-redis-client/tree/master/examples/clusters.php)
- [RAW Commands](https://github.com/cheprasov/php-redis-client/tree/master/examples/raw_commands.php)

## Installation

### Composer

Download composer:

    wget -nc http://getcomposer.org/composer.phar

and add dependency to your project:

    php composer.phar require cheprasov/php-redis-client

## Running tests

1. Run Docker container with Redis for tests https://hub.docker.com/r/cheprasov/redis-for-tests/
2. Run Docker container with Redis Cluster for tests https://hub.docker.com/r/cheprasov/redis-cluster-for-tests/
3. To run tests type in console:

    ./vendor/bin/phpunit

## Something doesn't work

Feel free to fork project, fix bugs and finally request for pull


CHANGELOG.md
## CHANGELOG

### v1.10.0 (2020-06-20)
- Added support Redis 6
- Updated server commands: ACL commands,
- Updated connection commands: CLIENT ID, CLIENT UNBLOCK, CLIENT CACHING, CLIENT GETREDIR, CLIENT TRACKING, HELLO
- Updated stream commands: XINFO (added FULL param)
- Added lists commands: LPOS
- Added strings commands: STRALGO LCS
- The client was checked with Redis-6.0.5, Redis-5.0.5, Redis-4.0.14 and older versions

### v1.9.1 (2019-10-12)
- Fixed bug: parser result error for transactions/pipelines https://github.com/cheprasov/php-redis-client/issues/72

### v1.9.0 (2019-07-16)
- Support Redis 5
- Added stream commands: XACK, XADD, XCLAIM, XDEL, XGROUP, XINFO, XLEN, XPENDING, XRANGE, XREAD, XREADGROUP, XREVRANGE, XTRIM.
- Added server commands: LOLWUT, REPLICAOF
- Added sorted sets commands: BZPOPMAX, BZPOPMIN, ZPOPMAX, ZPOPMIN
- The client was checked with Redis-5.0.5, Redis-4.0.14 and older versions

### v1.8.0 (2018-03-08)
- Added configuration for connection: timeout & flags.

### v1.7.2 (2017-08-19)
- Fixed bug of ClientFactory with default client version

### v1.7.0 (2017-07-22)
- The client was checked with Redis-4.0.0
- The client uses last stable redis version (Redis-4.0.0) by default.
- Deprecated client's methods: **executeRawString**, **parseRawString**
- Changed timeout on error **timeout_on_error_tryagain** from 0.25 sec to 0.05 sec.

### v1.6.1 (2017-02-04)
- Added check for empty data on reading response.
- Fixed some tests.

### v1.6.0 (2017-01-07)
- Added support for Redis 4.0 (the Client was tested with Redis 4.0 RC2).
- Added support for Redis Cluster.
- Added method **_syncClusterSlotsFromRedisServer** for RedisClient.
- Added command **SWAPDB**, **UNLINK**, **MEMORY** for Redis >= 4.0
- Updated command **FLUSHALL**, **FLUSHDB** for Redis >= 4.0

### v1.5.1 (2016-08-17)
- Fixed critical bug: https://github.com/cheprasov/php-redis-client/pull/45 Thanks to @BrianFranklin for help.

### v1.5.0 (2016-08-15)
- Added command **TOUCH** for Redis >= 3.2.1

### v1.4.0 (2016-07-18)
- You can choose default version of Redis Client (**ClientFactory::setDefaultRedisVersion**).
- Added parameter 'password' for config.
- Added parameter 'database' for config.

### v1.3.1 (2016-05-18)
- By default, the client works with the latest stable version of Redis (3.2.0).

### v1.3.0 (2016-05-07)
- Client was tested with Redis 3.2.0 (stable)
- Added command **BITFIELD** for Redis >= 3.2.
- Added **STORE** and **STOREDIST** params for **GEORADIUS** and **GEORADIUSBYMEMBER** for Redis >= 3.2.
- Added command **DEBUG HELP** for Redis >= 3.2.
- Changed some test for GEO.

### v1.2.3 (2016-05-02)
- Fixed command **PING** for Redis <= 2.6.
- Added common tests.

### v1.2.2 (2016-03-30)
- Fixed annotations and phpDocs.
- Fixed some tests.

### v1.2.1 (2016-03-04)
- Fixed some cluster commands (Redis >= 3.0): __CLUSTER COUNT-FAILURE-REPORTS__ and __CLUSTER SAVECONFIG__.

### v1.2.0 (2016-01-30)
- Updated command __MIGRATE__ (Redis >= 3.2): many fixes to the variable number of arguments (new) mode using the KEYS option.

### v1.1.0 (2016-01-26)
Sorry, no any history before.


examples/clusters.php
<?php
/**
 * This file is part of RedisClient.
 * git: https://github.com/cheprasov/php-redis-client
 *
 * (C) Alexander Cheprasov <acheprasov84@gmail.com>
 *
 * For the full copyright and license information, please view the LICENSE
 * file that was distributed with this source code.
 */

/**
 * Cluster
 */

namespace Examples;

require (dirname(__DIR__).'/vendor/autoload.php');
// or require (dirname(__DIR__).'/src/autoloader.php');

use RedisClient\ClientFactory;
use RedisClient\Cluster\ClusterMap;

// Example 1. Create RedisClient with Cluster support;
$RedisClient = ClientFactory::create([
    'server' => '127.0.0.1:7001', // Default server for connection
    'cluster' => [
        'enabled' => true
    ]
]);
print_r($RedisClient->clusterInfo());
/*
cluster_state:ok
cluster_slots_assigned:16384
cluster_slots_ok:16384
cluster_slots_pfail:0
cluster_slots_fail:0
cluster_known_nodes:6
cluster_size:3
cluster_current_epoch:6
cluster_my_epoch:1
cluster_stats_messages_sent:472694
cluster_stats_messages_received:472694
 */


// Example 2. Create RedisClient with Cluster Slot Map;
$RedisClient = ClientFactory::create([
    'server' => '127.0.0.1:7001', // Default server for connection
    'cluster' => [
        'enabled' => true,
        'clusters' => [
            5460  => '127.0.0.1:7001', // slots from 0 to 5460
            10922 => '127.0.0.1:7002', // slots from 5461 to 10922
            16383 => '127.0.0.1:7003', // slots from 10923 to 16383
        ],
    ]
]);


// Example 3. Get Cluster Slot Map from Redis Server on init RedisClient
$RedisClient = ClientFactory::create([
    'server' => '127.0.0.1:7001', // Default server for connection
    'cluster' => [
        'enabled' => true,
        // Use the param to update cluster slot map below on init RedisClient.
        // RedisClient will execute command CLUSTER SLOTS to get map.
        'init_on_start' => true,
    ]
]);


// Example 4. Get Cluster Slot Map from Redis Server by command
$RedisClient = ClientFactory::create([
    'server' => '127.0.0.1:7001', // Default server for connection
    'cluster' => [
        'enabled' => true,
    ]
]);
// Use this command to sync Cluster Slot Map from Redis Server
$RedisClient->_syncClusterSlotsFromRedisServer();


// Example 5. Pseudo Redis Cluster.
// If you use several Redis Servers without Cluster,
// you can use RedisClient's cluster config just for sharding keys by the Cluster rulers.
$RedisClient = ClientFactory::create([
    'server' => '127.0.0.1:6381', // Default server for connection
    'cluster' => [
        'enabled' => true,
        'clusters' => [
            5460  => '127.0.0.1:6381', // slots from 0 to 5460
            10922 => '127.0.0.1:6383', // slots from 5461 to 10922
            16383 => '127.0.0.1:6385', // slots from 10923 to 16383
        ],
    ]
]);
$RedisClient->set('foo', 'foo-42');
echo ClusterMap::getSlotByKey('foo') . PHP_EOL; // 12182

$RedisClient->set('bar', 'bar-42');
echo ClusterMap::getSlotByKey('bar') . PHP_EOL; // 5061

echo $RedisClient->get('foo') . PHP_EOL; // foo-42
echo $RedisClient->get('bar') . PHP_EOL; // bar-42

// But, be careful with multi operation for pseudo Clusters,
print_r($RedisClient->mget(['foo', 'bar'])); // ['foo-42', null]
print_r($RedisClient->mget(['bar', 'foo'])); // ['bar-42', null]


// Example 6. Each connection to Redis Server uses the same config.
// For example, use can use password for all servers
$RedisClient = ClientFactory::create([
    'server' => '127.0.0.1:6382', // Default server for connection
    'timeout' => 2,
    'password' => 'test-password-123',
    'cluster' => [
        'enabled' => true,
        'clusters' => [
            5460  => '127.0.0.1:6382', // slots from 0 to 5460
            10922 => '127.0.0.1:6384', // slots from 5461 to 10922
            16383 => '127.0.0.1:6386', // slots from 10923 to 16383
        ],
    ]
]);
$RedisClient->set('foo', 'foo-43');
echo ClusterMap::getSlotByKey('foo') . PHP_EOL; // 12182

$RedisClient->set('bar', 'bar-43');
echo ClusterMap::getSlotByKey('bar') . PHP_EOL; // 5061

echo $RedisClient->get('foo') . PHP_EOL; // foo-42
echo $RedisClient->get('bar') . PHP_EOL; // bar-42


